### ISSUE TITLE
Discount downloads from common CI environments

### ISSUE THREAD

**Issue by jsha** [rust-lang/crates.io#2886]
I was looking at the crate download code recently, and noticed that all downloads contribute to a crate's download count, even if it came from CI. While that's probably a useful number in terms of understanding how many resources a crate consumes from crates.io, it seems like download counts are commonly use to judge a package's popularity. If the counts include CI downloads, though, they reflect something more like "number of builds done by dependent projects" than "number of people / projects that depend on the crate."

Obviously it would be impossible to detect all possible CI environments, but three of the most popular set a specific environment variable: TRAVIS=true, CIRCLECI=true, and GITHUB_ACTIONS=true. Cargo could detect that it's being run in one of those environments and set a User-Agent header that includes "CI"; crates could then skip counting those downloads. This has some potential benefits in terms of reducing database load, since every counted download causes an UPDATE.

What do you think?

**Comment by Turbo87** [comment]
seems like a good idea to me on first glance 👍

**Comment by Turbo87** [comment]
@rust-lang/cargo is this something that you would want to handle on your side? if not, I guess we should close the issue

**Comment by ehuss** [comment]
The cargo team discussed this issue. In general it should be possible to provide more information, and we don't think there is necessarily any privacy concerns. 

However, we were uncertain about a few points:
* Would crates.io continue to track both "all downloads" and "all non-CI downloads"?
    * If separate, how would this information be presented differently?
* If this is primarily about popularity, why would discounting CI provide a better signal on popularity? If a crate is used frequently in CI, isn't that also a signal of popularity? 
    * If you care more about how many projects use a dependency, wouldn't a reverse dependency count be more accurate?
* If it just discounts CI, won't that skew the results in other ways (like other automated tooling)?

There were also some questions about the implementation (though not important to answer these now):
* Is tracking just `CI` environment variable enough? How many CI providers don't set that? Do they contribute enough traffic to matter?
* Should it instead be some kind of higher-level setting that the user could set to indicate "please turn off tracking"?
    * For example, a variety of tooling directly accesses the `static.crates.io` endpoint in part to avoid affecting download counts, but they are not "CI".
* If it is set in the user-agent, what happens if the user overrides the user-agent?
* Or should it use some other header or mechanism? Is it possible for the log collection to handle something besides user-agent?

**Comment by Turbo87** [comment]
@ehuss these are all very good questions and contrary to my initial comment on this issue a while ago, I'm no longer convinced that this will produce in the intended results. I've put this issue on the agenda for one of the next crates.io team meetings to check what the other team members think about this. unless we find evidence that this will result in significant advantages we might close this issue after all.

**Comment by Turbo87** [comment]
we discussed this in the crates.io team meeting today and came to the conclusion that we no long think discounting the CI downloads makes sense for us. we would be interested in additional metrics like "what percentage of downloads comes from CI systems", but we don't think that these should be discounted from the total number of downloads. filtering them out would be error-prone due to the difficulty of figuring out if a download actually came from CI or not, and so we would rather show a number that we are confident in, instead of a made up number that pretends to have filtered these systems out. additionally it would be hard to draw the line somewhere. e.g. would crates.io mirror downloads also be filtered out?

tl;dr not gonna happen ;)
